Add parameter named grains_blacklist

rongzeng54 opened the pull request:
### What does this PR do?
Add parameter named grains_blacklist to block the specific grains.

### What issues does this PR fix or reference?
No

### Previous Behavior
The default grains list, seems getting longer and uncontrolled by the user.

### New Behavior
The user finally has a chance to say no.
Of course, only for advanced users because of the dependencies of module.

### Tests written?
No

### Commits signed with GPG?
No

Commit:
Add parameter named grains_blacklist

cachedout commented:
Overall, I like the idea of this feature but I'm not certain about this implementation.

Would it not be more efficient to just filter at the point that the grains themselves are generated versus doing this on the instantiation of the LazyLoader? I wonder if we should move this logic to the `grains()` function in the loader. Let me know what you think, @rongzeng54

Commit:
move the logic of grains_blacklist to the grains function

rongzeng54 commented:
@cachedout 
Thank you for your reminder. Please check the new commit.

Commit:
fix for ret without anything

cachedout commented:
This is much better. I like this a lot.

cachedout commented:
@terminalmage You feel good about this?

terminalmage reviewed (changes requested):
This needs to have documentation added in `doc/ref/configuration/minion.rst`.

In addition, all other whitelists/blacklists that we support in salt use `salt.utils.stringutils.expr_match()` to support exact matches, `fnmatch.fnmatch()` matches, and regex matches. We should be using it here as well for consistency.

Lastly, rather than doing type checking, we should be setting a default value for this in `DEFAULT_MINION_OPTS` in `salt/config/__init__.py`. For example:

```python
'grains_blacklist': [],
```

There should also be an entry for this option in `VALID_OPTS` as well, with the valid type being `list`.

Commit:
changelogs:
- Update the documentation of minion configuration
- Use salt.utils.stringutils.expr_match() to support exact matches
- Setting a default value in DEFAULT_MINION_OPTS & VALID_OPTS

terminalmage commented on doc/ref/configuration/minion.rst:
```diff
@@ -752,6 +752,30 @@ Statically assigns grains to the minion.
+    Some modules may depends on core grains. Filtering may cause
+    the dependent module to run unusual.
+
+.. versionadded:: 2018.11.0
```
The version for this should be `Neon`. We haven't decided on a release number for the upcoming `Fluorine` release, but we've already passed the feature freeze for it anyway.

terminalmage commented on doc/ref/configuration/minion.rst:
```diff
@@ -752,6 +752,30 @@ Statically assigns grains to the minion.
+matched core grains will be filtered.
+
+.. note::
+
```
This blank line is unnecessary.

terminalmage commented on doc/ref/configuration/minion.rst:
```diff
@@ -752,6 +752,30 @@ Statically assigns grains to the minion.
+.. note::
+
+    Some modules may depends on core grains. Filtering may cause
+    the dependent module to run unusual.
```
This should be re-worded like this:

```
    Some states and execution modules depend on grains. Filtering may cause
    them to be unavailable or run unreliably.
```

terminalmage commented on doc/ref/configuration/minion.rst:
```diff
@@ -752,6 +752,30 @@ Statically assigns grains to the minion.
+.. conf_minion:: grains_blacklist
+
+``grains_blacklist``
+----------
```
The dashed line here needs to be the same length as the previous line to avoid a Sphinx warning.

terminalmage commented on salt/loader.py:
```diff
@@ -745,10 +742,12 @@ def grains(opts, force_refresh=False, proxy=None):
-            for bkey in bkeys:
-                del ret[bkey]
+        if blist:
+            for key in ret.keys():
```
This will break on Python 3:

```python
>>> x = {'foo': 'bar', 'abc': 123}
>>> for key in x.keys():
...     del x[key]
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: dictionary changed size during iteration
>>>
```

This is because on Python 3, `dict.keys()` is a view and not a separate list like in Python 2.

To fix this, use `list(ret)` instead of `ret.keys()`.

terminalmage commented on doc/ref/configuration/minion.rst:
```diff
@@ -752,6 +752,30 @@ Statically assigns grains to the minion.
+
+If :conf_minion:`grains_blacklist` is enabled then any core grains will be
+checked against the matches found in the `grains_blacklist` list and any
+matched core grains will be filtered.
```
We don't need to use the `:conf_minion:` Sphinx role here, linking to the same config option is redundant. We also should mention that globs and regex matches are both supported. How about this?

```
Each grains key will be compared against each of the expressions in this list.
Any keys which match will be filtered from the grains. Exact matches, glob
matches, and regular expressions are supported.
```

terminalmage commented on doc/ref/configuration/minion.rst:
```diff
@@ -752,6 +752,30 @@ Statically assigns grains to the minion.
+
+    grains_blacklist:
+      - cpu_flags
+      - zmq*
```
Since regexes are supported, a regex example should be added here as well.

terminalmage commented on salt/loader.py:
```diff
@@ -786,6 +785,14 @@ def grains(opts, force_refresh=False, proxy=None):
         if not isinstance(ret, dict):
             continue
+        if blist:
+            for key in ret.keys():
```
Same here, this should also be `list(ret)` instead of `ret.keys()`.

Commit:
changelogs:
- fix dict.keys() for python3
- fix the document error

rongzeng54 commented:
Hi @terminalmage 
Please check again.

isbm reviewed (commented):
Why this is even needed at all? If there is something in the default list, then it needs to be there _anyway_. If you want only specific keys, then you can already get them. But once you need something that you normally do not need and this "muter" exist, then you have a trouble.

I find this useless, sorry.

isbm commented:
> @isbm
> Useless data wastes server-side storage space, bandwidth caused by data synchronization, even speed.
> As described, It is not for everyone.

Then simply don't ask for all data. Ask only for what you need. But once you _do need_ it but at the same time _cannot get it_, then I find this a very bad idea.

rongzeng54 commented:
> > @isbm
> > Useless data wastes server-side storage space, bandwidth caused by data synchronization, even speed.
> > As described, It is not for everyone.
> 
> Then simply don't ask for all data. Ask only for what you need. But once you _do need_ it but at the same time _cannot get it_, then I find this a very bad idea.

The following is the configuration of our production environment, we believe that we cannot need them, but we also believe that others may need.
```
grains_blacklist:
  - 'lsb_distrib_codename'
  - 'lsb_distrib_id'
  - 'lsb_distrib_release'
  - 'cpu_flags'
  - 'fqdn_ip4'
  - 'fqdn_ip6'
  - 'num_gpus'
  - 'gpus'
  - 'locale_info'
  - 'machine_id'
  - 'pythonpath'
  - 'saltversioninfo'
  - 'pythonversion'
  - 'server_id'
  - 'saltpath'
  - 'domain'
  - 'localhost'
  - 'ip4_interfaces'
  - 'ip6_interfaces'
  - 'zmqversion'
  - 'SSDs'
  - 'disks'
  - 'mdadm'
  - 'username'
  - 'groupname'
  - 'uid'
  - 'gid'
  - 'systemd'
```

rongzeng54 commented:
Hi @terminalmage 
It seems that all the problems have been solved.

rallytime commented:
Can you write some tests for this new functionality? We wouldn't want this to regress in the future. :)

Commit:
add unittest for grains_blacklist

rongzeng54 commented:
> Can you write some tests for this new functionality? We wouldn't want this to regress in the future. :)

Hi @rallytime . It is done.

rallytime commented:
Thanks @rongzeng54!

max-arnold commented:
This option seems to partially duplicate the existing `disable_grains` parameter (briefly mentioned in #42567).

IMO, having two different ways to do this is confusing.

max-arnold commented:
@rongzeng54 I plan to open a PR to revert this change and instead document existing loader options (see the comment above). For example, you can prevent loading of expensive `esxi` grain module by adding `disable_grains: [esxi]` option.  This option is not exactly the same as yours in terms of granularity (it works on a module level), but it can have positive performance impact, see https://github.com/saltstack/salt/issues/48773#issuecomment-443847089

Do you have any objections against that?
